[#220] Add AWS Lambda / Serverless Framework guide and example code (#698)
slackapi/bolt-js: docs/_deployments/serverless-framework.md

---

> title: Serverless Framework

Author: In the side navation, I've chose to name the guide "Serverless Framework". The problem with this is that it doesn't jump out for someone looking for "AWS Lambda".

A few ideas:
1. Add a third side nav item called "AWS Lambda" that goes to this guide
1. Keep "Serverless Framework"
1. Rename to "AWS Lambda"

![image](https://user-images.githubusercontent.com/21328/100797167-8a39e500-33d6-11eb-8b21-bc61b8eb5cce.png)

Reviewer: Having the below list sounds easier to understand to me because serverless framework is not a platform but a tool.

* Heroku Platfform
* AWS Lambda

As the main topic here is about AWS and we use serverless framework as a solution for simpler deployments, I think we can rename it to "AWS Lambda".

Author: Thanks for your suggestion @seratch. I agree with you and I like the reasoning that Serverless Framework is a solution for simpler deploys, but we're deploying to AWS Lambda.

Author: In commit c8c8f42 I've updated our guide to be:
- Title: "Deploying to AWS Lambda"
- Path: `/aws-lambda`
- Side navigation: "AWS Lambda"
- Sample app: "Getting Started with AWS Lambda"

In the next commit, I'll rename:
- Guide filename: `aws-lambda.md`
- Sample app: `/examples/getting-started-with-aws-lambda`

order: 0
slug: serverless-framework
lang: en
layout: tutorial
permalink: /deployments/serverless-framework
redirect_from:
  - /deployments
---
# Deploying with the Serverless Framework

<div class="section-content">
This guide walks you through preparing and deploying a Slack app using [Bolt for JavaScript][bolt-js], the [Serverless Framework][serverless-framework], and [AWS Lambda][aws-lambda].
</div>

> When you’re finished, you’ll have this ⚡️[Getting Started with Serverless framework app][getting-started-with-serverless-framework-app] to run, modify, and make your own.

Author: This link will work after we merged this pull request. Until then you can view the app at:
https://github.com/mwbrooks/bolt-js/tree/220-faas-deployment-guides/examples/getting-started-with-serverless-framework

---

### Set up AWS Lambda

> [AWS Lambda][aws-lambda] is a serverless, function-as-a-service (FaaS) platform that allows you to run code without managing servers. In this section, we'll configure your local machine to access AWS Lambda.

Reviewer: ```suggestion
[AWS Lambda][aws-lambda] is a serverless, Function-as-a-Service (FaaS) platform that allows you to run code without managing servers. In this section, we'll configure your local machine to access AWS Lambda.
```

The quickest way to [configure your profile][aws-cli-configure] is to run this command and follow the prompts:

```zsh
aws configure
# AWS Access Key ID [None]: <your-aws-access-key>
# AWS Secret Access Key [None]: <your-aws-secret>
# Default region name [None]: us-east-1
# Default output format [None]: json
```

> 💡 Customize the [region][aws-cli-region] and [output format][aws-cli-output-format] best for you.

That wraps up configuring your local machine to access AWS. 👏 Next, let's do the same with the Serverless Framework.

---

### Set up Serverless Framework

Now we can set up the free [Serverless Framework](https://www.serverless.com/open-source/) open source tools on your local machine. These tools allow you to easily configure, debug, and deploy your app to AWS Lambda.

**1. Install the Serverless Framework CLI**

The Serverless tools are available as a Command Line Interface (CLI) and can be [installed on macOS, Windows, or Linux](https://www.serverless.com/framework/docs/getting-started/). You can install it using npm with the command:

```shell
npm install --save-dev serverless

> ```

Author: I've chosen to install the Serverless CLI as a developer dependency instead of globally. It's an approach that we used in [slackapi/bolt-examples-aws-re-invent-2020](https://github.com/slackapi/bolt-examples-aws-re-invent-2020/) and I like how this cuts down on the system installs that we ask of our community.

It also makes the CLI more portable for the other Bolt implementations, such as how we bundle the Serverless CLI with the Python example: https://github.com/slackapi/bolt-examples-aws-re-invent-2020/tree/main/api-gateway-lambda/python#4-deploy-to-aws-lambda

Thoughts?

Reviewer: Yeah, this makes sense.

**1. Prepare the app for AWS Lambda**

In this section, we'll update your Bolt app to respond to a Lambda function instead of listening for HTTP requests. We'll do this by creating a custom Bolt receiver for Lambda.

First, install the official [AWS Serverless Express](https://github.com/awslabs/aws-serverless-express) module to transform Express HTTP requests to Lambda function events:

```bash
npm install aws-serverless-express
```

Next, update the [source code that imports your modules](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L1) in `app.js` to require Bolt's Express receiver and the AWS Serverless Express module:

```javascript
const { App, ExpressReceiver } = require('@slack/bolt');
const awsServerlessExpress = require('aws-serverless-express');
```

Then update the [source code that initializes your Bolt app](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L3-L7) to create a custom receiver using AWS Serverless Express:

```javascript
// Initialize your custom receiver and ack Slack requests after handled
const expressReceiver = new ExpressReceiver({
  signingSecret: process.env.SLACK_SIGNING_SECRET,

>   processBeforeResponse: true

Reviewer: We may want to mention this a bit more clearly. This option is required for FaaS environments while it's not for other situations. Also, it is the only reason why we manually instantiate the receiver here.

Author: This is very good feedback, thanks @seratch. I'll take expand the comment to explain the importance and loop back with you!

Author: @seratch I've now added a detailed comment explaining the purpose of `processBeforeResponse`. What are your thoughts? Is it too length? Is the comment block above the option the ideal place? 

https://github.com/slackapi/bolt-js/pull/698/files#diff-05c75277f909db812ad95893db7558f993e8e51ff8f596273609b70ee2fa2ff9R132-R140

Reviewer: @mwbrooks Looks great 👍

});

// Initializes your app with your bot token and signing secret
const app = new App({
  token: process.env.SLACK_BOT_TOKEN,
  receiver: expressReceiver
});

// Initialize your AWSServerlessExpress server using Bolt's ExpressReceiver
const server = awsServerlessExpress.createServer(expressReceiver.app);
```

Finally, at the bottom of your app, update the [source code that starts the HTTP server](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L40-L45) to now respond to an AWS Lambda function event:

```javascript
// Handle the Lambda function event
module.exports.handler = (event, context) => {
  console.log('⚡️ Bolt app is running!');
  awsServerlessExpress.proxy(server, event, context);
};
```

> When you're done, your app should look similar to the ⚡️[Getting Started with Serverless Framework app][getting-started-with-serverless-framework-app/app.js].

Author: This link will work after the PR is merged, until then you can view the `app.js` at:
https://github.com/mwbrooks/bolt-js/blob/220-faas-deployment-guides/examples/getting-started-with-serverless-framework/app.js

**2. Update your Slack app's settings**

Now we need to use your AWS Lambda **endpoint** as your **Request URL**, which is where Slack will send events and actions.
Head over to your [Slack app's settings](https://api.slack.com/apps) and select your app name. We'll update your **Request URL** in two locations.

First, select **Interativity & Shortcuts** from the side and update the **Request URL**:

![Interactivity & Shortcuts page](../assets/interactivity-and-shortcuts-page.png "Interactivity & Shortcuts page")

Second, select **Event Subscriptions** from the side and update the **Request URL**:

![Event Subscriptions page](../assets/event-subscriptions-page.png "Event Subscriptions page")

**3. Test your Slack app**

Your app is now deployed and Slack is updated, so let's try it out!

Open a Slack channel that your app has joined and say "hello" (lower-case). Just like when we [ran the app locally](#run-the-app-locally), your app should respond back:

> 👩‍💻 hello<br/>
> 🤖 Hey there @Jane!

**4. Deploy an update**

As you continue building your Slack app, you'll need to deploy updates. Let's get a feel for this by updating your app to respond to a "goodbye" message.

> Add the following code to `app.js` ([source code on GitHub][getting-started-with-serverless-framework-app/app.js]):

Author: This link will work after the PR is merged, until then a preview of the link is at:
https://github.com/mwbrooks/bolt-js/blob/220-faas-deployment-guides/examples/getting-started-with-serverless-framework/app.js

```javascript
// Listens to incoming messages that contain "goodbye"
app.message('goodbye', async ({ message, say }) => {
  // say() sends a message to the channel where the event was triggered
  await say(`See ya later, <@${message.user}> :wave:`);
});
```

Deploy the update using the same command as before:

```shell
npx serverless deploy
```

When the deploy is complete, you can open a Slack channel that your app has joined and say "goodbye" (lower-case). You should see a friendly farewell from your Slack app.

---

### Next steps

> You just deployed your first ⚡️[Bolt for JavaScript app with Serverless to AWS Lambda][getting-started-with-serverless-framework-app]! 🚀

Author: Link will work after PR is merged, preview is at:
https://github.com/mwbrooks/bolt-js/tree/220-faas-deployment-guides/examples/getting-started-with-serverless-framework

---
title: Serverless Framework
order: 0
slug: serverless-framework
lang: en
layout: tutorial
permalink: /deployments/serverless-framework
redirect_from:
  - /deployments
---
# Deploying with the Serverless Framework

<div class="section-content">
This guide walks you through preparing and deploying a Slack app using [Bolt for JavaScript][bolt-js], the [Serverless Framework][serverless-framework], and [AWS Lambda][aws-lambda].
</div>

When you’re finished, you’ll have this ⚡️[Getting Started with Serverless framework app][getting-started-with-serverless-framework-app] to run, modify, and make your own.

---

### Set up AWS Lambda

[AWS Lambda][aws-lambda] is a serverless, Function-as-a-Service (FaaS) platform that allows you to run code without managing servers. In this section, we'll configure your local machine to access AWS Lambda.

> 💡 Skip this section if you have [configured a profile][aws-profiles] on your local machine with access to AWS Lambda.

**1. Sign up for an AWS account**

> First off, you'll need an AWS account. If you already an account, then you can go to the next step.

Author: Thanks, I like the suggestion to condense it as well. In commit 7b93ff435ad830c747605bf022e8526b40b91ace I've updated it to be:

```
If you don't already have an account, you should [sign up for AWS][aws-sign-up] and follow the on-screen instructions.
```

> First off, you'll need an AWS account. If you already an account, then you can go to the next step.

Author: Nice catch! From @shaydewael's feedback, we've updated this sentence to:

> If you don't already have an account, you should [sign up for AWS][aws-sign-up] and follow the on-screen instructions.

To create an account, [sign up for AWS](https://aws.amazon.com/) and follow the steps.

> 💡 You may be asked for payment information during the sign up. Don't worry, this guide only uses the [free tier][aws-pricing].

**2. Create an AWS access key**

Next, you'll need programmatic access to the AWS account to deploy onto AWS Lambda. In the world of AWS, this requires an **Access Key ID** and **Secret Access Key**.

We recommend watching this short, step-by-step video to 🍿 [create an IAM user and download the access keys](https://www.youtube.com/watch?v=KngM5bfpttA).

> > 💡 **Do you already have an IAM user?** Follow this official guide to [create access keys for an existing IAM user][aws-iam-user].

Author: Good catch! I've updated in commit 7b93ff435ad830c747605bf022e8526b40b91ace.

**3. Install the AWS CLI**

The AWS tools are available as a Command Line Interface (CLI) and can be [installed on macOS, Windows, or Linux][aws-cli-install].

On macOS, you can install the AWS CLI by [downloading the latest package installer](https://awscli.amazonaws.com/AWSCLIV2.pkg).

**4. Configure an AWS profile**

> You can use the AWS CLI to configure a profile that stores your access key pair on your local machine. This profile is used by the CLI and other tools to access AWS.

Author: Other tools include both first-party AWS tools (e.g. AWS SAM and AWS CDK) and third-party tools (e.g. Serverless, Apex, and Architect).

Do you think we should expand on the tools sentence or remove it? In the future, I imagine the AWS set up section becoming a partial that is used in other deployment guides that require AWS set up. So, I've tried to write this entire section agnostic to the Serverless Framework.

> You can use the AWS CLI to configure a profile that stores your access key pair on your local machine. This profile is used by the CLI and other tools to access AWS.

Reviewer 2: Good point. I like that it's agnostic. I think it's good to leave it

```shell
npm install --save-dev serverless
```

> 💡 Alternatively, you can [globally install the Serverless CLI](https://www.serverless.com/framework/docs/getting-started/) using the command `npm install -g serverless`.

Once the installation is complete, you can test the Serverless CLI by displaying the commands available to you:

```shell
npx serverless help
```

You're now set up with the Serverless tools! Let's move on to preparing your Bolt app to run as an AWS Lambda function.

---

### Get a Bolt Slack app

If you haven't already built your own Bolt app, you can use our [Getting Started guide][getting-started-guide] or clone the template app below:

```shell
git clone https://github.com/slackapi/bolt-js-getting-started-app.git
```

After you have a Bolt app, navigate to its directory:

```shell
cd bolt-js-getting-started-app/
```

Now that you have an app, let's prepare it for AWS Lambda and the Serverless Framework.

---

### Prepare the app

**1. Prepare the app for AWS Lambda**

> In this section, we'll update your Bolt app to respond to a Lambda function instead of listening for HTTP requests. We'll do this by creating a custom Bolt receiver for Lambda.

Reviewer 2: ```suggestion
By default, Bolt listens for HTTP requests. In this section, we'll customize your Bolt app's [`receiver`](https://slack.dev/bolt-js/concepts#receiver) to respond to Lambda function events instead.
```

Reviewer: >it'd be nice to mention the receiver concept up here

I agree 👍

```bash
npm install aws-serverless-express
```

Next, update the [source code that imports your modules](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L1) in `app.js` to require Bolt's Express receiver and the AWS Serverless Express module:

```javascript
const { App, ExpressReceiver } = require('@slack/bolt');
const awsServerlessExpress = require('aws-serverless-express');
```

Then update the [source code that initializes your Bolt app](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L3-L7) to create a custom receiver using AWS Serverless Express:

```javascript
// Initialize your custom receiver
const expressReceiver = new ExpressReceiver({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  // The `processBeforeResponse` option is required for all FaaS environments.
  // It allows Bolt methods (e.g. `app.message`) to handle a Slack request
  // before the Bolt framework responds to the request (e.g. `ack()`). This is
  // important because FaaS immediately terminate handlers after the response.
  processBeforeResponse: true
});

> // Initializes your app with your bot token and signing secret

Reviewer: Actually, we don't directly pass signing secret here. What do you think about changing like this (in a better language)?
```suggestion
// Initializes your app with your bot token and the AWS Lambda ready receiver
```

const app = new App({
  token: process.env.SLACK_BOT_TOKEN,
  receiver: expressReceiver
});

// Initialize your AWSServerlessExpress server using Bolt's ExpressReceiver
const server = awsServerlessExpress.createServer(expressReceiver.app);
```

Finally, at the bottom of your app, update the [source code that starts the HTTP server](https://github.com/slackapi/bolt-js-getting-started-app/blob/main/app.js#L40-L45) to now respond to an AWS Lambda function event:

```javascript
// Handle the Lambda function event
module.exports.handler = (event, context) => {
  console.log('⚡️ Bolt app is running!');
  awsServerlessExpress.proxy(server, event, context);
};
```

When you're done, your app should look similar to the ⚡️[Getting Started with Serverless Framework app][getting-started-with-serverless-framework-app/app.js].

**2. Add a serverless.yml**

> Every Serverless Framework app uses a special file called `serverless.yml` to configure and deploy your app.

Author: Personally, I'd lean toward "apps" since it keeps things simple by using common terminology in all Bolt guides. 

Otherwise, we may need to do a deeper dive into the Serverless configuration -  at the moment, we don't even explain the details of the `serverless.yml` configuration.

Create a new file called `serverless.yml` in your app's root directory and paste the following:

```yaml
service: serverless-bolt-js
frameworkVersion: '2'
provider:
  name: aws
  runtime: nodejs12.x
  environment:
    SLACK_SIGNING_SECRET: ${env:SLACK_SIGNING_SECRET}
    SLACK_BOT_TOKEN: ${env:SLACK_BOT_TOKEN}
functions:
  slack:
    handler: app.handler
    events:
      - http:
          path: slack/events
          method: post
plugins:
  - serverless-offline
```

> 💡 `SLACK_SIGNING_SECRET` and `SLACK_BOT_TOKEN` must be enviornment variables on your local machine.
> You can [learn how to export Slack environment variables](/bolt-js/tutorial/getting-started#setting-up-your-local-project) in our Getting Started guide.

**3. Install Serverless Offline** 

To make local development a breeze, we can use the `serverless-offline` module to emulate a deployed function.

Run the following command to install it as a development dependency:

```bash
npm install --save-dev serverless-offline
```

Congratulations, you've just prepared your Bolt app for AWS Lambda and Serverless! Now we can learn how to run and deploy the app.

---

### Run the app locally

> Let's learn how to run your Bolt app on your local machine, now that it's configured to respond to an AWS Lambda function.

Author: @shaydewael Is this suggestion missing part of the sentence?

> Let's learn how to run your Bolt app on your local machine, now that it's configured to respond to an AWS Lambda function.

Reviewer 2: Oops yep, it was missing part of it. Modified it, but you can ignore the suggestion if you don't agree with it. I think flipping the sentence structure makes sense though.

> Let's learn how to run your Bolt app on your local machine, now that it's configured to respond to an AWS Lambda function.

Author: This is much better, thanks! I've made the change in commit 70c71099d4efe3ccf078c44b8596207989cd1a6c

**1. Start your local servers**

First, use the Serverless Framework to listen for an AWS Lambda function event:

```zsh
npx serverless offline --noPrependStageInUrl
```

Next, use ngrok to forward Slack events to your local machine:

```zsh
ngrok http 3000
```

> > 💡 [Learn how to install ngrok](https://api.slack.com/tutorials/tunneling-with-ngrok) to create a public URL and forward requests to your local machine.

Author: Thanks! When I wrote this section, I debated referencing the install tutorial or Getting Started guide. I leaned on the tutorial because it was more focused, but I'll update it to the Getting Started guide because it adds more context!

> > 💡 [Learn how to install ngrok](https://api.slack.com/tutorials/tunneling-with-ngrok) to create a public URL and forward requests to your local machine.

Author: Updated in commit 7b93ff435ad830c747605bf022e8526b40b91ace to be:

> 💡 [Learn how to use ngrok][getting-started-guide-setting-up-events] to create a public URL and forward requests to your local machine

First, select **Interativity & Shortcuts** from the side and update the **Request URL**:

![Interactivity & Shortcuts page](../assets/interactivity-and-shortcuts-page.png "Interactivity & Shortcuts page")

Second, select **Event Subscriptions** from the side and update the **Request URL**:

![Event Subscriptions page](../assets/event-subscriptions-page.png "Event Subscriptions page")

**3. Test your Slack app**

Now we can test your Slack app by opening a Slack channel that your app has joined and saying “hello” (lower-case). Just like in the [Getting Started guide][getting-started-guide], your app should respond back:

> 👩‍💻 hello<br/>
> 🤖 Hey there @Jane!

If you don’t receive a response, check your **Request URL** and try again.

> 💡 **How does this work?**
> The ngrok and Serverless commands are configured on the same port (default: 3000). When a Slack event is sent to your **Request URL**, it's received on your local machine by ngrok. The request is then forwarded to Serverless Offline, which emulates an AWS Lambda function event and triggers your Bolt app's receiver. 🛫🛬 Phew, what a trip!

---

### Deploy the app

> We'll use the Serverless Framework tools to provision, package, and deploy your app onto AWS Lambda. After your app is deployed, we'll update your Slack app's settings, and say "hello" to your app. ✨

Author: Thanks! I'll commit this suggestion, because it's better than the original then try to update the wording.
